### DISCUSSION TITLE
About application rules on Clean Architecture

### DISCUSSION THREAD

**Discussion by diogokobbi** [techexcellenceio/techexcellence#12] [Q&A]
Hello, i have a doubt about the Application layer on Clean Architecture Diagram. Our entities from domain layer have business rules, so we have a central point to validate operations, but what about the application layer? should we have the same behaviour there for specific uses cases?

Let me give a example: here in Brazil our banking systems do not allow some operations on mobile applications, only on desktop apps. So, how do we apply our rules to block or hide some info for specific apps? Where theses rules should belong to?

I tought that uses case methods of each application (mobile and desktop) should be responsible for that, but some examples available on github show only one application (like UserService, for example) in the application layer. I'm talking about something like this:

// 2 projects on application layer:
- (1) MobileAppplicationService
  -  ShowBalance()
    // On this use case implementation, we hide (block or format) entity data to display?
- (2) DesktopApplicationService
  -  ShowBalance()
  // display all info

// 2 projects on external layer::
- (1) AndroidApp
  // calls WebApi project that depends on MobileAppplicationService
- (2) Windows Desktop App
  // calls WebApi project that depends on DesktopAppplicationService

Maybe my example is not clear enough. A food delivery app can have the same issue: restaurant and client share the same Order Entity, but each one has different views and operations on presenter layer. Where should those specific rules for each presenter belong to?

If we attach business logic in the presenter layer, we cannot evolve a restaurant webform application to React SPA without lose or rewrite application rules, for example.

**Comment by valentinajemuovic** [comment]
Good question. Use cases should be independent of the presentation layer. We could have any number of presentation devices but they should not impact the ApplicationService / UseCase.

Going to your example: The ApplicationService class (or UseCase class) should be independent of the UI, so we don't use the word Mobile/Desktop in the name, nor anywhere else

So the following are two valid options I see (whereby Use Cases remain decoupled from Presentation)

- Option 1: Mobile App & Desktop App would calling the same Web API, then Web Api calls the ApplicationService. Mobile App shows only some of the data from the response, whereas the Desktop App shows all the data from the response.
- Option 2: Backend for Frontend (BFF). We'd have MobileWebApi and DesktopWebApi. They would both have an endpoint show-balance BUT the difference is that the MobileWebApi is returning a response with just some data and the DesktopWebApi is returning a response with all the data about the balance.

In this way, the presentation layer (Dekstop, Mobile, REST API) is sensitive to presentation needs... BUT the application layer does not know anything about Mobile/Desktop.
